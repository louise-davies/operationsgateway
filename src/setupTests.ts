// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import crypto from 'crypto';
// TODO JOEL: Is this still needed? - Also in vite config
// need to mock <canvas> for plotting
import 'vitest-canvas-mock';
// TODO JOEL: Replace with vitest-fail-on-console
import failOnConsole from 'jest-fail-on-console';
import { TextEncoder } from 'util';
import { server } from './mocks/server';

// TODO JOEL: Is this still needed? - same for failOnConsole below
global.TextEncoder = TextEncoder;

failOnConsole();

vi.setConfig({ testTimeout: 15000 });

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());

// TODO JOEL: Check if a all of the below is still needed

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {},
  });
}

// this is needed because of https://github.com/facebook/jest/issues/8987
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let mockActualReact;
vi.doMock('react', async () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (!mockActualReact) {
    mockActualReact = await vi.importActual('react');
  }
  return mockActualReact;
});

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.revokeObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {
      // no-op
    },
  });
}

// jest doesn't implement web crypto so set up nodejs crypto as a default
Object.defineProperty(global, 'crypto', {
  value: Object.setPrototypeOf({ subtle: crypto.subtle }, crypto),
});
