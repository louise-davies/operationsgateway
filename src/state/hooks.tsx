import { useDispatch, useSelector } from 'react-redux';
import type { TypedUseSelectorHook } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { createSelector } from '@reduxjs/toolkit';

// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

const everythingButConfigSelector = ({ config, ...state }: RootState) => state;
export const sessionSelector = createSelector(
  everythingButConfigSelector,
  (state) => state,
  {
    devModeChecks: {
      identityFunctionCheck: 'never',
      inputStabilityCheck: 'never',
    },
  }
);
