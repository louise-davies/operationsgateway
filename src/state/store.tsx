import { configureStore } from '@reduxjs/toolkit';
import OperationsGatewayMiddleware, {
  listenToMessages,
} from './middleware/operationsgateway.middleware';
import configReducer from './slices/configSlice';

export const store = configureStore({
  reducer: {
    config: configReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(OperationsGatewayMiddleware),
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

listenToMessages(store.dispatch);
