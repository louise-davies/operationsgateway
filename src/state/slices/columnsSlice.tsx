import { createSelector, createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import { Column } from 'react-table';
import { DropResult } from 'react-beautiful-dnd';
import { RootState } from '../store';
import { Order } from '../../app.types';
import { resultsPerPage } from '../../recordGeneration';

// Define a type for the slice state
interface ColumnsState {
  columnDefs: {
    [id: string]: Column;
  };
  selectedColumnIds: string[];
  page: number;
  resultsPerPage: number;
  sort: {
    [column: string]: Order;
  };
}

// Define the initial state using that type
export const initialState = {
  columnDefs: {},
  selectedColumnIds: [],
  page: 0,
  resultsPerPage: resultsPerPage,
  sort: {},
} as ColumnsState;

export const columnsSlice = createSlice({
  name: 'columns',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    // Use the PayloadAction type to declare the contents of `action.payload`
    setColumns: (state, action: PayloadAction<Column[]>) => {
      // reduce the collection by the id property into a shape of { 1: { ...user }}
      const byId = action.payload.reduce(
        (
          byId: {
            [id: string]: Column;
          },
          col
        ) => {
          byId[col.accessor?.toString() ?? ''] = col;
          return byId;
        },
        {}
      );
      state.columnDefs = byId;
    },
    // Use the PayloadAction type to declare the contents of `action.payload`
    setSelectedColumns: (state, action: PayloadAction<Column[]>) => {
      state.selectedColumnIds = action.payload.map(
        (col) => col.accessor?.toString() ?? ''
      );
    },
    selectColumn: (state, action: PayloadAction<string>) => {
      state.selectedColumnIds.push(action.payload);
    },
    deselectColumn: (state, action: PayloadAction<string>) => {
      delete state.sort[action.payload];

      const newSelectedColumnsIds = state.selectedColumnIds.filter(
        (colId) => colId !== action.payload
      );
      state.selectedColumnIds = newSelectedColumnsIds;
    },
    reorderColumn: (state, action: PayloadAction<DropResult>) => {
      const result = action.payload;
      if (result.destination) {
        const [reorderedItem] = state.selectedColumnIds.splice(
          result.source.index,
          1
        );
        state.selectedColumnIds.splice(
          result.destination?.index,
          0,
          reorderedItem
        );
      }
    },
    changeSort: (
      state,
      action: PayloadAction<{ column: string; order: Order | null }>
    ) => {
      const { column, order } = action.payload;
      if (order !== null) {
        state.sort[column] = order;
      } else {
        delete state.sort[column];
      }
    },
    changePage: (state, action: PayloadAction<number>) => {
      state.page = action.payload;
    },
    changeResultsPerPage: (state, action: PayloadAction<number>) => {
      state.resultsPerPage = action.payload;
      state.page = 0;
    },
  },
});

export const {
  setColumns,
  setSelectedColumns,
  selectColumn,
  deselectColumn,
  reorderColumn,
  changeSort,
  changePage,
  changeResultsPerPage,
} = columnsSlice.actions;

// Other code such as selectors can use the imported `RootState` type
export const selectColumnDefs = (state: RootState) => state.columns.columnDefs;
export const selectSelectedIds = (state: RootState) =>
  state.columns.selectedColumnIds;
export const selectSort = (state: RootState) => state.columns.sort;
export const selectPage = (state: RootState) => state.columns.page;
export const selectResultsPerPage = (state: RootState) =>
  state.columns.resultsPerPage;

function arrayEquals(a: string[], b: string[]) {
  const sortedA = [...a].sort();
  const sortedB = [...b].sort();
  return (
    Array.isArray(sortedA) &&
    Array.isArray(sortedB) &&
    sortedA.length === sortedB.length &&
    sortedA.every((val, index) => val === sortedB[index])
  );
}

const selectSelectedIdsIgnoreOrder = createSelector(
  selectSelectedIds,
  (selectedIds) => selectedIds,
  {
    memoizeOptions: { equalityCheck: arrayEquals },
  }
);

export const selectSelectedColumns = createSelector(
  selectColumnDefs,
  selectSelectedIdsIgnoreOrder,
  (columnDefs, selectedIds) => {
    return Object.values(columnDefs).filter((col) => {
      if (col.accessor?.toString()) {
        return selectedIds.includes(col.accessor.toString());
      } else {
        return false;
      }
    });
  }
);

export const selectHiddenColumns = createSelector(
  selectColumnDefs,
  selectSelectedIdsIgnoreOrder,
  (columnDefs, selectedIds) => {
    return Object.keys(columnDefs).filter((colId) => {
      return !selectedIds.includes(colId);
    });
  }
);

export const selectAvailableColumns = createSelector(
  selectColumnDefs,
  (columnDefs) => {
    return Object.values(columnDefs);
  }
);

export default columnsSlice.reducer;
