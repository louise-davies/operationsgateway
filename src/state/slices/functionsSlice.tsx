import type { PayloadAction } from '@reduxjs/toolkit';
import { createSelector, createSlice } from '@reduxjs/toolkit';
import { convertExpressionsToStrings } from '../../api/functions';
import { ValidateFunctionState } from '../../app.types';
import { RootState } from '../store';

// Define a type for the slice state
interface FunctionsState {
  appliedFunctions: ValidateFunctionState[];
}

// Define the initial state using that type
export const initialState = {
  appliedFunctions: [],
} as FunctionsState;

export const functionsSlice = createSlice({
  name: 'functions',
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    // Use the PayloadAction type to declare the contents of `action.payload`
    changeAppliedFunctions: (
      state,
      action: PayloadAction<ValidateFunctionState[]>
    ) => {
      state.appliedFunctions = action.payload;
    },
  },
});

export const { changeAppliedFunctions } = functionsSlice.actions;

// Other code such as selectors can use the imported `RootState` type
export const selectAppliedFunctions = (state: RootState) =>
  state.functions.appliedFunctions;

export const selectQueryFunctions = createSelector(
  selectAppliedFunctions,
  (appliedFunctions) =>
    JSON.parse(convertExpressionsToStrings(appliedFunctions))
);

export default functionsSlice.reducer;
