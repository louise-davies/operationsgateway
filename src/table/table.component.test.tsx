import React from 'react';
import Table, { TableProps } from './table.component';
import { screen, cleanup, act, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { RecordRow } from '../app.types';
import { Column } from 'react-table';
import { flushPromises, renderWithProviders, getState } from '../setupTests';
import { RootState } from '../state/store';
import { PreloadedState } from '@reduxjs/toolkit';
import { changeSort, selectColumn } from '../state/slices/columnsSlice';

describe('Table', () => {
  let props: TableProps;
  const recordRows: RecordRow[] = [
    {
      timestamp: new Date('2022-01-01T00:00:00Z').getTime().toString(),
      activeArea: '1',
      shotNum: 1,
      activeExperiment: '1',
    },
    {
      timestamp: new Date('2022-01-02T00:00:00Z').getTime().toString(),
      activeArea: '2',
      shotNum: 2,
      activeExperiment: '2',
    },
    {
      timestamp: new Date('2022-01-03T00:00:00Z').getTime().toString(),
      activeArea: '3',
      shotNum: 3,
      activeExperiment: '3',
    },
  ];
  const columnDefs: { [id: string]: Column } = {
    timestamp: {
      Header: 'Timestamp',
      accessor: 'timestamp',
    },
    shotNum: {
      Header: 'Shot Number',
      accessor: 'shotNum',
    },
    activeArea: {
      Header: 'Active Area',
      accessor: 'activeArea',
    },
    activeExperiment: {
      Header: 'Active Experiment',
      accessor: 'activeExperiment',
    },
  };
  const onPageChange = jest.fn();
  const onResultsPerPageChange = jest.fn();
  const onSort = jest.fn();
  let state: PreloadedState<RootState>;

  const createView = (initialState = state) => {
    return renderWithProviders(<Table {...props} />, {
      preloadedState: initialState,
    });
  };

  beforeEach(() => {
    props = {
      data: recordRows,
      totalDataCount: recordRows.length,
      page: 0,
      loadedData: true,
      loadedCount: true,
      resultsPerPage: 25,
      onPageChange: onPageChange,
      onResultsPerPageChange: onResultsPerPageChange,
      onSort: onSort,
      sort: {},
    };
    state = { columns: { ...getState().columns, columnDefs } };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly, with timestamp column displayed by default', () => {
    const view = createView();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders correctly with all columns displayed', async () => {
    state.columns.selectedColumnIds = [
      'timestamp',
      'shotNum',
      'activeArea',
      'activeExperiment',
    ];
    const view = createView();

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('displays a record count', () => {
    const recordCount = recordRows.length;
    createView();
    screen.getByText(`1–${recordCount} of ${recordCount}`);
  });

  it('calls onPageChange when page is changed', async () => {
    const user = userEvent.setup();
    const recordCount = recordRows.length;
    props.resultsPerPage = 1;
    createView();
    screen.getByText(`1–1 of ${recordCount}`);

    await user.click(screen.getByLabelText('Go to next page'));

    await act(async () => {
      await flushPromises();
    });

    expect(onPageChange).toHaveBeenCalledWith(1);
  });

  // TODO: improve test once fixed to pagination errors is merged
  it('calls onResultsPerPageChange when resultsPerPage is changed', async () => {
    const user = userEvent.setup();
    createView();
    const resultsPerPage = screen.getByRole('button', {
      name: /Rows per page/i,
    });
    await user.click(resultsPerPage);

    const listbox = within(screen.getByRole('listbox'));

    await user.click(listbox.getByText('10'));

    await act(async () => {
      await flushPromises();
    });

    expect(onResultsPerPageChange).toHaveBeenCalledWith(10);
  });

  it('displays page loading message when loadedData is false and totalDataCount is zero', async () => {
    props.loadedData = false;
    createView();
    screen.getByRole('progressbar');

    cleanup();
    props.loadedData = true;
    props.totalDataCount = 0;
    createView();
    screen.getByRole('progressbar');
  });
});
