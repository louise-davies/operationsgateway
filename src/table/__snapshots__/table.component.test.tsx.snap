// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
        >
          <table
            class="MuiTable-root css-rqglhn-MuiTable-root"
            role="table"
            style="min-width: 90px;"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                role="row"
                style="display: flex; flex: 1 0 auto; min-width: 90px;"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1fvzch-MuiTableCell-root"
                  role="columnheader"
                  scope="col"
                >
                  <div
                    style="overflow: hidden; flex: 1;"
                  >
                    <div
                      aria-label="id header"
                      class="MuiBox-root css-k008qs"
                    >
                      <div
                        class="MuiBox-root css-12z0wuy"
                      />
                      <div
                        class="MuiBox-root css-0"
                      >
                        <span
                          class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                          data-testid="sort id"
                          role="button"
                          tabindex="0"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-xw0cpc-MuiTypography-root"
                          >
                            ID
                          </p>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                            data-testid="ArrowDownwardIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                            />
                          </svg>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    style="display: flex; flex-direction: row;"
                  >
                    <div
                      aria-label="close id"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1pmilg0-MuiSvgIcon-root"
                        data-testid="CloseIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                    </div>
                    <div
                      draggable="false"
                      role="separator"
                      style="margin-left: 10px; padding-left: 4px; cursor: col-resize;"
                    >
                      <hr
                        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-ck7sc0-MuiDivider-root"
                      />
                    </div>
                  </div>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1fvzch-MuiTableCell-root"
                  role="columnheader"
                  scope="col"
                >
                  <div
                    style="overflow: hidden; flex: 1;"
                  >
                    <div
                      aria-label="shotNum header"
                      class="MuiBox-root css-k008qs"
                    >
                      <div
                        class="MuiBox-root css-12z0wuy"
                      />
                      <div
                        class="MuiBox-root css-0"
                      >
                        <span
                          class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                          data-testid="sort shotNum"
                          role="button"
                          tabindex="0"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-xw0cpc-MuiTypography-root"
                          >
                            Shot Number
                          </p>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                            data-testid="ArrowDownwardIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                            />
                          </svg>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    style="display: flex; flex-direction: row;"
                  >
                    <div
                      aria-label="close shotNum"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1pmilg0-MuiSvgIcon-root"
                        data-testid="CloseIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                    </div>
                    <div
                      draggable="false"
                      role="separator"
                      style="margin-left: 10px; padding-left: 4px; cursor: col-resize;"
                    >
                      <hr
                        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-ck7sc0-MuiDivider-root"
                      />
                    </div>
                  </div>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall css-1fvzch-MuiTableCell-root"
                  role="columnheader"
                  scope="col"
                >
                  <div
                    style="overflow: hidden; flex: 1;"
                  >
                    <div
                      aria-label="timestamp header"
                      class="MuiBox-root css-k008qs"
                    >
                      <div
                        class="MuiBox-root css-12z0wuy"
                      />
                      <div
                        class="MuiBox-root css-0"
                      >
                        <span
                          class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                          data-testid="sort timestamp"
                          role="button"
                          tabindex="0"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-xw0cpc-MuiTypography-root"
                          >
                            Timestamp
                          </p>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                            data-testid="ArrowDownwardIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                            />
                          </svg>
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    style="display: flex; flex-direction: row;"
                  >
                    <div
                      aria-label="close timestamp"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1pmilg0-MuiSvgIcon-root"
                        data-testid="CloseIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                    </div>
                    <div
                      draggable="false"
                      role="separator"
                      style="margin-left: 10px; padding-left: 4px; cursor: col-resize;"
                    >
                      <hr
                        class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-ck7sc0-MuiDivider-root"
                      />
                    </div>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
              role="rowgroup"
            >
              <tr
                class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                role="row"
                style="display: flex; flex: 1 0 auto; min-width: 90px;"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        test1
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        1
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        1640995200000
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
              </tr>
              <tr
                class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                role="row"
                style="display: flex; flex: 1 0 auto; min-width: 90px;"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        test2
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        2
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        1641081600000
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
              </tr>
              <tr
                class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                role="row"
                style="display: flex; flex: 1 0 auto; min-width: 90px;"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        test3
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        3
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-1aeaoy8-MuiTableCell-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body2 MuiTypography-noWrap css-z1doim-MuiTypography-root"
                      >
                        1641168000000
                      </p>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
                    />
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
        >
          <div
            class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
          />
          <p
            class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
            id=":r1:"
          >
            Rows per page:
          </p>
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
            variant="standard"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby=":r1: :r0:"
              class="MuiTablePagination-select MuiSelect-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
              id=":r0:"
              role="button"
              tabindex="0"
            >
              10
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="10"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTablePagination-selectIcon MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
          <p
            class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
          >
            1–3 of 3
          </p>
          <div
            class="MuiTablePagination-actions"
          >
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
              disabled=""
              tabindex="-1"
              title="Go to previous page"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowLeftIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                />
              </svg>
            </button>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
              disabled=""
              tabindex="-1"
              title="Go to next page"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowRightIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
